### DISCUSSION TITLE
is_reachable:"unknown"

### DISCUSSION THREAD

**Discussion by ewhent** [reacherhq/check-if-email-exists#1376] [Q&A]
Can I access a list of all unique messages generated when the "is_reachable" field is marked as "unknown" in your tool? If not, do you have recommendations for efficiently capturing and categorizing these messages?

**Comment by amaury1093** [comment]
AFAIK there's no exhaustive list of all messages. I started trying to categorize them into 2 categories:
- unknown because IP is blacklisted ([code](https://github.com/reacherhq/check-if-email-exists/blob/5ca4dfa5ec38fba0ec7cfb052106da8d6af4df44/core/src/smtp/parser.rs#L148))
- unknown because server needs a rDNS ([code](https://github.com/reacherhq/check-if-email-exists/blob/5ca4dfa5ec38fba0ec7cfb052106da8d6af4df44/core/src/smtp/parser.rs#L218))

There might be other categories to group `unknown` messages, but I didn't explore them fully yet.

If the `unknown` message matches one of these 2 categories, the `smtp` field in the JSON output will describe the category `IpBlacklisted` or `NeedsRDNS`. If not, then the unknown message will end up uncategorized.

**Reply by ewhent** [reply]
I've been testing the tool with various emails, and I noticed that for "unknown" statuses, the categorization into "IpBlacklisted" or "NeedsRDNS" isn't always present in the smtp field.

Instead, I observed an "error" field within parentheses, including "type" and "message." Could you provide more insights into how these "error" field types are intended to be interpreted, especially when the categorization into "IpBlacklisted" or "NeedsRDNS" is not present?

**Reply by amaury1093** [reply]
Yes, `type` and `message` are always present. Reacher then parses these 2 fields, and when they match something familiar, Reacher _additionally_ adds a `IpBlacklisted` or `NeedsRDNS` category.

However, there are many other unknown `type`/`message` combinations that are not familiar. They require a human looking through them individually.

If someone looks at a large amount of those `type`/`message` combinations, we can potentially find a pattern (e.g. multiple similar messages). Once such pattern is found, I (or anyone else via a PR!) can add it to Reacher's pattern-matching algorithm.

**Reply by ewhent** [reply]
I appreciate your prompt response. I've been testing the tool on a set of emails that consistently generate "is_reachable" as "unknown." According to your explanation, there's a possibility that some of these should be categorized as "IpBlacklisted" or "NeedsRDNS."

However, upon reviewing the JSON output for one of the tested emails (see below), I couldn't find any indication of these specific categories.

json:

[
    {
        "input": "XYZ@EMAIL.COM",
        "is_reachable": "unknown",
        "misc": {
            "is_disposable": false,
            "is_role_account": false
        },
        "mx": {
            "accepts_mail": true,
            "records": [
                "eu-smtp-inbound-1.mimecast.com.",
                "eu-smtp-inbound-2.mimecast.com."
            ]
        },
        "smtp": {
            "error": {
                "type": "SmtpError",
                "message": "transient: Internal resource temporarily unavailable - https://community.mimecast.com/docs/DOC-1369#451 [8QxuRhpcPp-1YEIRp4XrSA.uk201]"
            }
        },
        "syntax": {
            "address": "XYZ@EMAIL.COM",
            "domain": "EMAIL.COM",
            "is_valid_syntax": true,
            "username": "XYZ"
        }
    }
]
Could you provide further insights into why this particular case didn't receive the "IpBlacklisted" or "NeedsRDNS" categorization? I want to ensure I'm interpreting the output correctly and contribute effectively to the tool's improvement.

Thank you for your assistance.

**Reply by amaury1093** [reply]
In your case, as per the given link https://community.mimecast.com/docs/DOC-1369#451, the root cause is this:

<img width="813" alt="Screenshot 2023-11-17 at 08 40 39" src="https://github.com/reacherhq/check-if-email-exists/assets/1293565/85ca181f-128e-4be8-93d1-19e2a5e907c5">

It's definitely not `NeedsRDNS`. It's not really `IpBlacklisted`. I could create a 3rd category `IpGreylisted`.

**Reply by amaury1093** [reply]
> The introduction of a new category, "IpGreylisted," emphasizes the dynamic nature of categorizations in real-world scenarios.
> 
> This prompts the realization that there may be additional categories beyond the existing ones like "IpBlacklisted" or "NeedsRDNS." It highlights the evolving landscape of email verification and the need for human intervention to discern and categorize unknown scenarios effectively.
> 
> In essence, the understanding is that we may not be certain of all potential categories generated when the "is_reachable" field is marked as "unknown," underscoring the importance of ongoing human involvement for accurate interpretation.
> 
> If there are further considerations or specific steps you recommend for community involvement in this process, I'd be eager to contribute.

Why are you pasting a ChatGPT response?

**Reply by ewhent** [reply]
I am just reframing my texts just so that I sound appropriate and correct 

"So now this is a third category and there can be more than this right? so basically we arent sure of all the categories that can be generated when "is_reachable" field is "unknown" so it needs human intervention."

this is what I wanted to convey but just for making it sound correct I reframed it
